compose/foundation/foundation/src/desktopMain/kotlin/androidx/compose/foundation/text/ContextMenu.desktop.kt

```diff
@@ -66,59 +81,204 @@ internal fun OpenMenuAdjuster(state: ContextMenuState, adjustAction: (Offset) ->
+         */
+        val cut: (() -> Unit)?
+
+        /**
+         * Action for copy the selected text to the clipboard. Null if there is no text to copy.
+         */
+        val copy: (() -> Unit)?
+
+        /**
+         * Action for pasting text from the clipboard. Null if there is no text in the clipboard.
+         */
+        val paste: (() -> Unit)?
+
+        /**
+         * Action for selecting the whole text. Null if the text is already selected.
+         */
+        val selectAll: (() -> Unit)?
+    }
+
+    companion object {
+        /**
+         * [TextContextMenu] that is used by default in Compose.
+         */
+        @ExperimentalFoundationApi
+        val Default = object : TextContextMenu {
+            @Composable
+            override fun Area(actions: Actions, state: ContextMenuState, content: @Composable () -> Unit) {
+                val localization = LocalLocalization.current
+                val items = {
+                    listOfNotNull(
+                        actions.cut?.let {
+                            ContextMenuItem(localization.cut, it)
+                        },
+                        actions.copy?.let {
+                            ContextMenuItem(localization.copy, it)
+                        },
+                        actions.paste?.let {
+                            ContextMenuItem(localization.paste, it)
+                        },
+                        actions.selectAll?.let {
+                            ContextMenuItem(localization.selectAll, it)
+                        },
+                    )
                 }
-            )
+
+                ContextMenuArea(items, state, content = content)
+            }
         }
-        result
     }
 }
 
-@Composable
-internal fun SelectionManager.contextMenuItems(): () -> List<ContextMenuItem> {
-    val localization = LocalLocalization.current
-    return {
-        listOf(
-            ContextMenuItem(localization.copy) {
-                copy()
+/**
+ * [TextContextMenu] that uses [JPopupMenu] to show the text context menu.
+ *
+ * You can use it by overriding [TextContextMenu] on the top level of your application.
+ *
+ * @param owner The root component that owns a context menu. Usually it is [ComposeWindow] or [ComposePanel].
+ * @param createMenu Describes how to create [JPopupMenu]. Use it if you want customization of the menu.
+ * @param createItem Describes how to create a generic context menu item. Use it if you want customization.
+ * of the menu items. It is called for items provided by [ContextMenuArea] in user code.
+ * @param createCut Describes hot to create the menu item for Cut action.
+ * @param createCopy Describes hot to create the menu item for Copy action.
+ * @param createPaste Describes hot to create the menu item for Paste action.
+ * @param createSelectAll Describes hot to create the menu item for Select All action.
+ */
+@ExperimentalFoundationApi
+class JPopupTextMenu(
```

Comment:
If users need to have more menu items (not only cur,copy,paste,selectAll), then they can't use JPopupTextMenu and they need to create their own TextContextMenu, right?

Reply:
If I want to create a JPopupMenu with my custom items, I need to:

- extend Actions interface
- implement TextContextMenu and cast Actions to my own extended Actions in `fun Area` implementation

so it would be quite a copy/paste of some JPopupTextMenu parts, right?

Reply:
Good points!

It seems we don't have to pass all items below text context menu, because usually it is a final destination, and we don't need to provide item further down the tree. And because of that, I also removed `ContextManuDataProvider` from `ContextMenuArea` and `distinctBy`.

That fact allowed to significantly simplify interfaces, and now we can create more customizable menus (with separator, for example)

It also fixes this [issue](https://github.com/JetBrains/compose-jb/issues/979) the more right way (but now it has a crash)